### ISSUE TITLE
on reward definition on sentiment control example

### ISSUE THREAD

**Issue by DaehanKim** [lvwerra/trl#120]
Thank you for this awesome repo! I really like this project. 😀 

while reading through tutorial notebooks, I wondered why you defined the neutral reward as `-2 * pos_logit + 4`.
Also, as I think, defining positive reward as `pos_logit` and negative reward `-pos_logit` may not be a good choice.

Since Pipeline outputs pre-softmax logits, resulting probability values depend on the difference between positive and negative logits, not the absolute values. So I think it's better to have `pos_reward = pos_logit - neg_logit` and `neg_reward = neg_logit - pos_logit`. 

For the neutral reward, `abs(pos_logit - neg_logit)` should be minimized to get a response that most confuses the sentiment pipeline. so `-abs(pos_logit - neg_logit)` might be a good choice. 

Am I missing something or is there another reason you define rewards that way?

**Comment by DaehanKim** [comment]
Thank you for answering. 
I'm running the sentiment-control example with changed reward definition. will report when finished!

**Comment by DaehanKim** [comment]
Changed reward shows better reward plot. 

![image](https://user-images.githubusercontent.com/20675681/216758541-7ce89ee4-1806-440c-b01c-6d4e6d989451.png)

Changed reward is defined as follow: 
```
def logits_to_reward(logits, task):
    """
    Take the positive and negative logits and scale it for the task.
        task [negative]: reward = neg_logit - pos_logit
        task [neutral]: reward = -abs(pos_logit - neg_logit) + 4
        task [positive]: reward = pos_logit - neg_logit
    logits : List of tensors (negative_logit, positive_logit)
    """python3
    rewards = list()
    for i in range(len(logits)):
        if task[i]=='[negative]':
            rewards.append(logits[i][0] - logits[i][1])
        elif task[i]=='[neutral]':
            rewards.append(-torch.abs(logits[i][0] - logits[i][1]) + 4)
        elif task[i]=='[positive]':
            rewards.append(logits[i][1] - logits[i][0])
        else:
            raise ValueError('task has to be in [0, 1, 2]!')
    return rewards
```

I also tried step reward definition. 
If [positive] it gives reward 1 if `pos_logit - neg_logit > 1`
If [negative] it gives reward 1 if `neg_logit - pos_logit > 1`
If [neutral] it gives reward 1 if `abs(pos_logit - neg_logit) < 0.4` (This results in less than 0.1 probability difference btw pos and neg)

This doesn't work and is even worse than original implementation.

Two runs stopped because of sentiment pipeline error. It accidently received >512 token inputs. I suspect this happens because of non-english characters that explodes sentiment token counts. I'll take a look and report when I figure it out.

**Comment by lvwerra** [comment]
Thanks for the report, this looks great! The issue with the loss spikes is now fixed with #126. One thing to keep in mind is that since you are changing the reward definition the reward curve's scale also changes. E.g. the new negative reward has twice the range of the original one.

**Comment by DaehanKim** [comment]
@lvwerra That's a good point I missed out!
I'll run with `0.5*new_reward` and see how they are different. Thanks!

**Comment by DaehanKim** [comment]
I ran experiments with new scales and things doesn't seem quite improved. So I suppose reward is well-defined empirically. (@lvwerra 👍 ) 

![image](https://user-images.githubusercontent.com/20675681/218448042-6ed0c644-5ee9-49c3-bad7-c49c5cdac2ad.png)

[full wandb charts here](https://wandb.ai/lucas01/trl?workspace=user-lucas01)

changed reward was defined as follow: 
```
def logits_to_reward(logits, task):
    """
    Take the positive sentiment logit and scale it for the task.
        task [negative]: reward = -(pos_logit - neg_logit)*0.5
        task [neutral]: reward = -abs(pos_logit - neg_logit)+4
        task [positive]: reward = (pos_logit - neg_logit)*0.5
    """
    rewards = []
    for i in range(len(logits)):
        if task[i]=='[negative]':
            rewards.append(0.5*(logits[i][0] - logits[i][1]))
        elif task[i]=='[neutral]':
            rewards.append(-torch.abs(logits[i][0] - logits[i][1]) + 4)
        elif task[i]=='[positive]':
            rewards.append(0.5*(logits[i][1] - logits[i][0]))
        else:
            raise ValueError('task has to be in [0, 1, 2]!')
    return rewards
```

**Comment by DaehanKim** [comment]
Thanks man! I suppose it's good to close the issue. Please reopen when necessary!
